modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadLaws.kt

```diff
@@ -117,4 +133,39 @@ object MonadLaws {
     forAll(GK.genK(Gen.int()), GK.genK(Gen.int())) { fa, fb ->
       AP.run { fa.apTap(fb) }.equalUnderTheLaw(fa.apTap(fb), EQ)
     }
+
+  fun <F> Monad<F>.derivedFlattenConsistent(GK: GenK<F>, EQ: Eq<Kind<F, Int>>): Unit =
+    forAll(GK.genK(GK.genK(Gen.int()))) { fa: Kind<F, Kind<F, Int>> ->
+      fa.flatten().equalUnderTheLaw(fa.flatMap(::identity), EQ)
+    }
+
+  fun <F> Monad<F>.derivedFollowedByEvalConsistent(GK: GenK<F>, AP: Applicative<F>, EQ: Eq<Kind<F, Int>>): Unit =
+    forAll(GK.genK(Gen.int()), Gen.int().eval(AP)) { fa, fb: Eval<Kind<F, Int>> ->
+      fa.followedByEval(fb).equalUnderTheLaw(fa.flatMap { fb.value() }, EQ)
+    }
+
+  fun <F> Monad<F>.derivedProductLConsistent(GK: GenK<F>, EQ: Eq<Kind<F, Int>>): Unit =
+    forAll(GK.genK(Gen.int()), GK.genK(Gen.int())) { fa, fb ->
+      fa.productL(fb).equalUnderTheLaw(fa.flatMap { a -> fb.map { a } }, EQ)
+    }
+
+  fun <F> Monad<F>.derivedProductLEvalConsistent(GK: GenK<F>, AP: Applicative<F>, EQ: Eq<Kind<F, Int>>): Unit =
+    forAll(GK.genK(Gen.int()), Gen.int().eval(AP)) { fa, fb: Eval<Kind<F, Int>> ->
+      fa.productLEval(fb).equalUnderTheLaw(fa.flatMap { a -> fb.value().map { a } }, EQ)
+    }
+
+  fun <F> Monad<F>.derivedProductConsistent(GK: GenK<F>, AP: Apply<F>, EQ: Eq<Kind<F, Tuple2<Int, Int>>>): Unit =
+    forAll(GK.genK(Gen.int()), GK.genK(Gen.int())) { fa, fb ->
+      AP.run { fa.product(fb) }.equalUnderTheLaw(fa.product(fb), EQ)
+    }
+
+  fun <F> Monad<F>.derivedMProductConsistent(GK: GenK<F>, EQ: Eq<Kind<F, Tuple2<Int, Int>>>): Unit =
+    forAll(GK.genK(Gen.int()), Gen.functionAToB<Int, Kind<F, Int>>(Gen.int().applicative(this))) { fa, fb: (Int) -> Kind<F, Int> ->
+      fa.mproduct(fb).equalUnderTheLaw(fa.flatMap { a -> fb(a).map { Tuple2(a, it) } }, EQ)
+    }
+
+  fun <F> Monad<F>.derivedIfMConsistent(GK: GenK<F>, EQ: Eq<Kind<F, Int>>): Unit =
+    forAll(GK.genK(Gen.bool()), Gen.functionToA(Gen.int().applicative(this)), Gen.functionToA(Gen.int().applicative(this))) { fa, fTrue, fFalse ->
+      fa.ifM(fTrue, fFalse).equalUnderTheLaw(fa.flatMap { if (it) fTrue() else fFalse() }, EQ)
+    }
```

Comment:
Please don't use `Gen<A>.applicative()` unless you really want a type to only be `just(a)`. This will for example never generate the `None` case of `Option` or `Left` of `Either`. Prefer the `GenK.genK(genA)` version as that properly handles those cases.

Reply:
Oh and the same applies to the `Gen.eval` generator you added, please use `GenK` instead, it provides much better test-data.

Reply:
Should we get rid of `Gen<A>.applicative()` in that case? Or deprecate it for ourselves?

Reply:
@nomisRev I think it depends, you might need gens of pure liftings in certain situations. In a recent PR i'm actually adding one explicit for raising errors only

Reply:
That behaviour is indeed useful in some situations, but I'd prefer to get rid of it and write an explicit `genA.map { just(it) }` instead. Since most tests already have the instance in scope that should be as easy as writing `genA.applicative(this)`.
Another option is to rename the generator because `applicative` is easy to misunderstand. Something like `Gen.pure`, `Gen.just` etc might be better to explicitly show what this generator produces. Plus a note that points out the weaknesses and refers to `GenK`.

Reply:
Thank you @1Jajen1 for the explanation!! I've removed eval and replaced applicative in the tests